src/pac.c

```diff
@@ -105,47 +108,65 @@ pac_may_have_muzzy(pac_t *pac) {
 }
 
 static edata_t *
-pac_alloc_real(tsdn_t *tsdn, pac_t *pac, ehooks_t *ehooks, size_t size,
-    size_t alignment, bool zero, bool guarded) {
-	assert(!guarded || alignment <= PAGE);
+pac_alloc_unguarded(tsdn_t *tsdn, pac_t *pac, size_t size, size_t alignment,
+    bool zero, bool *deferred_work_generated) {
+	ehooks_t *ehooks = pac_ehooks_get(pac);
 
 	edata_t *edata = ecache_alloc(tsdn, pac, ehooks, &pac->ecache_dirty,
-	    NULL, size, alignment, zero, guarded);
+	    NULL, size, alignment, zero, false);
 
 	if (edata == NULL && pac_may_have_muzzy(pac)) {
 		edata = ecache_alloc(tsdn, pac, ehooks, &pac->ecache_muzzy,
-		    NULL, size, alignment, zero, guarded);
+		    NULL, size, alignment, zero, false);
 	}
 	if (edata == NULL) {
 		edata = ecache_alloc_grow(tsdn, pac, ehooks,
-		    &pac->ecache_retained, NULL, size, alignment, zero,
-		    guarded);
+		    &pac->ecache_retained, NULL, size, alignment, zero);
 		if (config_stats && edata != NULL) {
 			atomic_fetch_add_zu(&pac->stats->pac_mapped, size,
 			    ATOMIC_RELAXED);
 		}
 	}
 
 	return edata;
+
 }
 
 static edata_t *
-pac_alloc_new_guarded(tsdn_t *tsdn, pac_t *pac, ehooks_t *ehooks, size_t size,
-    size_t alignment, bool zero) {
-	assert(alignment <= PAGE);
-
-	size_t size_with_guards = size + SAN_PAGE_GUARDS_SIZE;
-	/* Alloc a non-guarded extent first.*/
-	edata_t *edata = pac_alloc_real(tsdn, pac, ehooks, size_with_guards,
-	    /* alignment */ PAGE, zero, /* guarded */ false);
-	if (edata != NULL) {
-		/* Add guards around it. */
-		assert(edata_size_get(edata) == size_with_guards);
-		san_guard_pages(tsdn, ehooks, edata, pac->emap, true, true,
-		    true);
-	}
-	assert(edata == NULL || (edata_guarded_get(edata) &&
-	    edata_size_get(edata) == size));
+pac_alloc_guarded(tsdn_t *tsdn, pac_t *pac, size_t size, size_t alignment,
+    bool zero, bool *deferred_work_generated) {
+	ehooks_t *ehooks = pac_ehooks_get(pac);
+
+	edata_t *edata;
+	edata = ecache_alloc(tsdn, pac, ehooks, &pac->ecache_dirty, NULL, size,
+	    alignment, zero, true);
+
+	if (edata == NULL && pac_may_have_muzzy(pac)) {
+		edata = ecache_alloc(tsdn, pac, ehooks, &pac->ecache_muzzy,
+		    NULL, size, alignment, zero, true);
+	}
+	if (edata == NULL) {
+		edata = ecache_alloc(tsdn, pac, ehooks, &pac->ecache_retained,
+		    NULL, size, alignment, zero, true);
+
+		if (edata != NULL && config_prof) {
+			extent_gdump_add(tsdn, edata);
+		}
+	}
+	if (size >= SC_LARGE_MINCLASS || !opt_retain) {
+		edata = pac_alloc_unguarded(tsdn, pac,
```

Comment:
Discussed offline - `bool slab` flag will be added to PAI to pass this data down to `PAC`.

Reply:
Can you go into some more detail? It was a lot of work for the page-level allocator not to know about the purpose for which its pages is used, and it's fairly abstraction breaking to do so. We should try to do something more specific if at all possible.

Reply:
Agreed that we shouldn't pass down another param really mean "is_slab". What we want is actually a hint, more like `bool frequent_reuse_expected`, which doesn't have to be exactly slab (but we may simply use the slab as the criteria). The PA will then use it to determine if a guarded extent should be cached (prevents coalescing but speeds up reuse), or getting rid of the guards (mprotect needed).

The other thought is, I think it's also worth trying out some hints with the HPA later, e.g. to group slabs into the same hugepage.

An alternative to the additional hint param, is to have a size threshold within the PA, to indicate that, e.g. sizes <= MAX_SLAB_SIZE are frequently reused. It may get a bit awkward when the dynamic slab sizes are used though (then a large range of sizes may get covered).

Reply:
Feels compelling to me, thanks.